### PR TITLE
Add OCaml 4.11 support.

### PR DESCRIPTION
**Pull request by Et7f3**
What changed:
- `Pconst_string` has now a new location attribute (middle)
- `Typedtree.case` changed to `Typedtree.case(computation)` or `Typedtree.case(value)`. So another level of nesting + it is declared private thus forcing to write 2 nested match. Also most annotation `: Typedtree.case` aren't compatible between version. The prefix field access work because field name are different hopefully.
- `let _: unit => 'a = fail;` is too ignore fail because we won't before < 4.11
- ` type k. Compat.collectPattern(k)` k is used only in 4.11
- OCaml >= 4.08 has a filter_map (thus allowing to use double match + assert(false)

TODO:
- [x] List.filter_map (texpMatchGetExceptions)
- [ ] remove annotation ? (getTexpMatch, typedCaseCont, cCases) added primarly for debug can be removed now

Why Arnold ? :joy:

I have seen a `filter_map` in `dead-code-ppx/DeadCodePPX.re` which one should be used ? The one present or the one in stdlib ?

### CODE REVIEW

**Comment by cristianoc** [comment]
@Et7f3 thanks a lot for doing this!

Quick Q: is there a way to avoid the `type k` to leak outside the `Compat` module? E.g. `let collectPattern: type k. Compat.collectPattern(k) = (super, self, pat) => {` in `DeadValue.re`.

**Comment by Et7f3** [comment]
Why it is a issue ? Is is a phantom type on <4.11

**Comment by cristianoc** [comment]
Not a huge deal but it raises the barrier to contribution, as it's something most people don't know how to use, and is only there because of 4.11.

**Comment by Et7f3** [comment]
I can move cppo `#if` of lazyload-ppx/LazyLoad.re to Compate.re. I can also add a comment to explain to lower the barrier.

**Comment by cristianoc** [comment]
Btw LazyLoad is only used on 4.06, so one could even if-def it completely away on 4.11.

**Comment by Et7f3** [comment]
ok now lazyload.exe is builded only on 4.06 :white_check_mark:
and since I have bumped dune lang you get many cool feature: like opam generation, formatting (which I haven't used to not make review harder: `dune build @fmt --auto-promote`), enabled_if (used in this PR), ...

**Comment by cristianoc** [comment]
I've tested the changes in CI here: https://github.com/reason-association/reanalyze/pull/98/.
Avoiding changes to dune so CI is not disturbed.

I wanted to test on 4.11, but here's what I get:
```sh
% opam install reason

The following dependencies couldn't be met:
  - reason → ocaml < 4.11
```

How did you get to build on 4.11?

**Comment by Et7f3** [comment]
Oops I have used resolution with esy and thus bypassed constraint.
Add this in esy.json:
```json
{
  "source": "./reanalyze.opam",
  "override": {
    "dependencies": {
      "@opam/reason": null,
      "@esy-ocaml/reason": "3.6.2"
    },
    "devDependencies": {
      "@opam/reason": null
    }
  }
}```

**Comment by Et7f3** [comment]
here in 98 you reverted part of my initial commit. during git blame this will force to blame 2 times. Do you want I update my commit then ?

**Comment by cristianoc** [comment]
Still playing with the changes a bit, to see how to minimize the amount of specific logic going into compat.

**Comment by cristianoc** [comment]
I think this can be wrapped up.
I can either merge https://github.com/reason-association/reanalyze/pull/98, or let me know if you'd prefer to create a new PR with a single commit.

**Comment by Et7f3** [comment]
Ok I will do a cleaning. So are you fine with dune 2.3 bump (dune 2.7.1 is the latest) and since dune >= 2 the dev experience is much better.

**Comment by cristianoc** [comment]
Better try any changes to dune/opam files later in a separate PR, to check separately how CI is affected. Changes have created problems in the past.

**Comment by cristianoc** [comment]
Thanks for updating the PR!
Is it ready to merge?

**Comment by Et7f3** [comment]
Yes

File: src/Compat.re [hunk 1]
```diff
@@ -7,6 +7,63 @@ let getStringTag = s => switch (s) {
...
+#else
+type generalPattern('a) = Typedtree.pattern
+#endif
+
+let unboxPatCstrName = (pat) => {
+#if OCAML_MINOR >= 11
+  switch (pat) {
+    | Typedtree.Tpat_value(v) =>
+      switch ((v :> Typedtree.pattern_data(Typedtree.pattern_desc(Typedtree.value))).pat_desc) {
+        | Tpat_construct(_, {cstr_name}, _) => Some(cstr_name)
```
Comment (Et7f3) [inline src/Compat.re:54]: I think I have a tricks to reduce code duplication: we ifdef only on a branch the we recursively call with just the Tpat_construct hence Tpat_construct is not duplicated.
Need to test and if Tpat construct is changed in future versions this trick might not work anymore so don't know if you want it.

**Review by Et7f3 — COMMENTED** [review]
If I haven't committed when you see it again: then you can merge it.

File: src/Compat.re (hunk 1 above)
Comment (Et7f3) [reply src/Compat.re:54]: if you want to apply `patch -p1 < recu.txt`

Comment (Et7f3) [reply src/Compat.re:54]: [recu.txt](https://github.com/reason-association/reanalyze/files/5240205/recu.txt)
If you want to apply `patch -p1 < recu.txt`

Comment (Et7f3) [reply src/Compat.re:54]: [recu.txt](https://github.com/reason-association/reanalyze/files/5240205/recu.txt)
If you want to apply `patch -p1 < recu.txt`
I'm not convinced it simply so not pushed but I share it FWIW.

**Comment by cristianoc** [comment]
Merging.

**Pull request merged**